spec/index.bs

```diff
@@ -1033,17 +1033,48 @@ its mapping.
      mapping.
 
 <script type=idl>
+typedef [EnforceRange] unsigned long GPUMapUsageFlags;
+interface GPUMapUsage {
+    const GPUMapUsageFlags READ  = 0x0001;
+    const GPUMapUsageFlags WRITE = 0x0002;
+    const GPUMapUsageFlags READ_WRITE = 0x0003;
+};
+
+interface GPUMappedSubrange {
+    ArrayBuffer mapped;
+};
+
 [Serializable]
 interface GPUBuffer {
-    Promise<ArrayBuffer> mapReadAsync();
-    Promise<ArrayBuffer> mapWriteAsync();
-    void unmap();
+    Promise<void> mapRangeAsync(GPUBufferMapFlags flags, unsigned long long offset = 0, unsigned long long size = 0);
+    GPUMappedSubrange? getMappedSubrange(unsigned long long offset = 0, unsigned long long size = 0);
+    void unmapRange(unsigned long long offset = 0, unsigned long long size = 0);
 
     void destroy();
 };
 GPUBuffer includes GPUObjectBase;
 </script>
 
+If size is zero, {{GPUBuffer/map}} treats it as an effective size of buffer-length-starting-at-offset.
+Effective size of a mapping must not be zero.
+{{GPUBuffer/map}} fails (and rejects) if the range overlaps with already-mapped ranges in GPUBuffer.
+{{GPUBuffer/map}}'s Promise resolves before {{GPUFence/onCompletion}}'s Promise for any
+{{GPUQueue/signal}} enqueued after the last {{GPUQueue/submit}} that uses the buffer.
+If |mapRangeAsync| requests a subrange that is not in the "unmapped" state, the promise rejects.
+(If two duplicate calls to |mapRangeAsync| are made back-to-back, the second will always reject)
+When the promise from |mapRangeAsync| resolves, its subrange is marked as "mapped".
+
+If |getMappedRange| is called while the requested subrange is not in the "mapped" state,
+|getMappedRange| fails and returns null.
+If |getMappedRange| succeeds, the subrange is set to the "visible" state.
+
+It is an error if |unmapRange| specifies range that:
+* Is not "mapped" or "visible".
```

Comment:
does a range even have a single "state" in this approach? If I map 0..10 but then 0..4 become "visible", then the whole 0..10 doesn't have a single state that we can talk about